src/mod/endpoints/mod_sofia/mod_sofia.c

```diff
@@ -3967,13 +3967,19 @@ static switch_status_t cmd_profile(char **argv, int argc, switch_stream_handle_t
 			stream->write_function(stream, "+OK\n");
 		} else if ((gateway_ptr = sofia_reg_find_gateway(gname))) {
 			if (gateway_ptr->state != REG_STATE_NOREG && gateway_ptr->state != REG_STATE_DOWN) {
-				gateway_ptr->retry = 0;
-				gateway_ptr->state = REG_STATE_UNREGISTER;
+				if (gateway_ptr->state == REG_STATE_FAILED) {
```

Comment:
I don't get the logic here, so currently when the gateway is neither NOREG or DOWN state, we will just set it to UNREGISTER which will send again a REGISTER with expire then transition to DOWN state. When the current state was FAILED then it will just mark it as UNREGISTER then it will transition to DOWN state again. The transition should trigger firing the gateway state event. What is the purpose for this code changed?

Reply:
Also when it mark as FAILED state it should automatically transition to DOWN state or FAIL_WAIT state. The only time it will be in the FAILED state is during timeout, then it will transition afterwards to FAIL_WAIT > DOWN state. If for example the RE-REGISTER timeouts, if we follow this logic then it will no longer attempt to UNREGISTER it and mark it as DOWN state. In KSS the record will still be in the database since we didn't sent the UNREGISTER request

Reply:
As discussed in the meeting

`sofia_reg_handle_sip_r_register` will set the FAILED status for a failed registration and trigger the event. That happens not only on timeout. It happens when receiving a negative response from KSS for INITIAL REGISTRATION and REREGISTRATION

Then `on_sofia_gateway_state` will catch the event and trigger a `sofia profile internal unregister <GW>`

Then we arrive to this part of the code mentioned. If the FAILED registration was the initial registration, FS will send an UNREGISTER towards KSS for a GW that was never registered. This piece of code avoids that.

With this change we won't send the UNREGISTER towards KSS if what fails is a RE-REGISTER,.

As agreed, I will add a new reg_flag to mark if the gateway was ever registered or not. Then I can modify this piece to check for the flag: if it was previously registered, send the UNREGISTER. if not, mark the GW as DOWN and fire the event